### PR TITLE
ExcavatorNHMP (Excavator version 1.5.5a for Nicehash only)

### PR DESCRIPTION
**Pull request by UselessGuru**
@Aaron: I do not own multiple Nvidia cards, so I cannot test multiple algo/worker combinations.

### CODE REVIEW

**Comment by Dewbs** [comment]
For me Pools/NiceHash.ps1 causes Nicehash to stop mining (current fixed nicehash.ps1 works fine)
ExcavatorNicehash.ps1 never downloads the bin

File: MultiPoolMiner.ps1 [hunk 1]
```diff
@@ -298,8 +298,8 @@ while ($true) {
...
     # since mining is probably still working.  Then it filters out any algorithms that aren't being used.
     $AllPools = @($NewPools) + @(Compare-Object @($NewPools | Select-Object -ExpandProperty Name -Unique) @($AllPools | Select-Object -ExpandProperty Name -Unique) | Where-Object SideIndicator -EQ "=>" | Select-Object -ExpandProperty InputObject | ForEach-Object {$AllPools | Where-Object Name -EQ $_}) | 
-        Where-Object {$Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm $_.Algorithm -IncludeEqual -ExcludeDifferent | Measure-Object).Count -gt 0} | 
-        Where-Object {$Config.ExcludeAlgorithm.Count -eq 0 -or (Compare-Object $Config.ExcludeAlgorithm $_.Algorithm -IncludeEqual -ExcludeDifferent | Measure-Object).Count -eq 0} | 
+        Where-Object {$Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm ($_.Algorithm  -split "-" | Select-Object -Index 0) -IncludeEqual -ExcludeDifferent | Measure-Object).Count -gt 0} | 
+        Where-Object {$Config.ExcludeAlgorithm.Count -eq 0 -or (Compare-Object $Config.ExcludeAlgorithm ($_.Algorithm -split "-" | Select-Object -Index 0) -IncludeEqual -ExcludeDifferent | Measure-Object).Count -eq 0} |
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: Without this the suffixed algorithms (*-NHMP) would get filtered out -> No pools for ExcavatorNHMP
Algorithm_Norm will never output anything cointaining '-' so this is safe.

The alternative would be to extend the algorithm list in start.bat

File: MultiPoolMiner.ps1 [hunk 2]
```diff
@@ -346,8 +346,8 @@ while ($true) {
...
             Where-Object {(Compare-Object @($Devices.Name | Select-Object) @($_.DeviceName | Select-Object) | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0} | 
-            Where-Object {($Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0) -and ((Compare-Object $Pools.PSObject.Properties.Name $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0)} | 
-            Where-Object {$Config.ExcludeAlgorithm.Count -eq 0 -or (Compare-Object $Config.ExcludeAlgorithm $_.HashRates.PSObject.Properties.Name -IncludeEqual -ExcludeDifferent | Measure-Object).Count -eq 0} | 
+            Where-Object {($Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm ($_.HashRates.PSObject.Properties.Name -split "-" | Select-Object -Index 0) | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0) -and ((Compare-Object $Pools.PSObject.Properties.Name $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0)} |
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: Without this the suffixed algorithms (*-NHMP) would get filtered out -> No miners for ExcavatorNHMP
Algorithm_Norm will never output anything cointaining '-' so this is safe.

The alternative would be to extend the algorithm list in start.bat

File: MultiPoolMiner.ps1 [hunk 3]
```diff
@@ -346,8 +346,8 @@ while ($true) {
...
-            Where-Object {$Config.ExcludeAlgorithm.Count -eq 0 -or (Compare-Object $Config.ExcludeAlgorithm $_.HashRates.PSObject.Properties.Name -IncludeEqual -ExcludeDifferent | Measure-Object).Count -eq 0} | 
+            Where-Object {($Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm ($_.HashRates.PSObject.Properties.Name -split "-" | Select-Object -Index 0) | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0) -and ((Compare-Object $Pools.PSObject.Properties.Name $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0)} | 
+            Where-Object {$Config.ExcludeAlgorithm.Count -eq 0 -or (Compare-Object $Config.ExcludeAlgorithm ($_.HashRates.PSObject.Properties.Name -split "-" | Select-Object -Index 0) -IncludeEqual -ExcludeDifferent | Measure-Object).Count -eq 0} |
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: Without this the suffixed algorithms (*-NHMP) would get filtered out -> No miners for ExcavatorNHMP
Algorithm_Norm will never output anything cointaining '-' so this is safe.

The alternative would be to extend the algorithm list in start.bat

File: MultiPoolMiner.ps1 [hunk 4]
```diff
@@ -421,7 +421,7 @@ while ($true) {
         $Miner.Path = $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Miner.Path)
         if ($Miner.PrerequisitePath) {$Miner.PrerequisitePath = $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Miner.PrerequisitePath)}
 
-        if ($Miner.Arguments -isnot [String]) {$Miner.Arguments = $Miner.Arguments | ConvertTo-Json -Compress}
+        if ($Miner.Arguments -isnot [String]) {$Miner.Arguments = $Miner.Arguments | ConvertTo-Json -Depth 10 -Compress}
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: No more truncated JSON statements

**Comment by UselessGuru** [comment]
@Dewbs 

> causes Nicehash to stop mining

Fixed. Was caused by an invalid nicehash host definition in pool file.

> never downloads the bin

This is a follow-up error cause by the affected pool file.

**Comment by Dewbs** [comment]
Tried and it does download and mine but I think the benchmark result is incorrect.
It lists miner in each gpu section but then assigns collective rate to each gpu.
It's like the miners are split per gpu but the result is recorded as if they are combined (to all).

**Comment by Kettoch** [comment]
Testing this change now.

**Comment by Kettoch** [comment]
I get the error when merging these files into mine.
`WARNING: Pool prices are out of sync (22 minutes).
Compare-Object : Cannot bind argument to parameter 'ReferenceObject' because it is null.
At C:\MultiPoolMiner\MultiPoolMiner.ps1:532 char:129
+ ... pare-Object ($_.Combination | Select-Object -ExpandProperty DeviceNam ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidData: (:) [Compare-Object], ParameterBindingValidationException
+ FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.CompareObjectCommand`

**Comment by Kettoch** [comment]
I Believe this API call is wrong.  Each one has a double up of this.

api | Received: {"id":1,"method":"worker.add","params":["equihash","0","equihash","1"]}


Their API Refence gives the list of additional parameters.  It may be that MPM is trying to add more than one worker at a time and this API does not support it.

https://github.com/nicehash/excavator/tree/master/nvidia

**Comment by Kettoch** [comment]
Seems to be something to do with the API Call, 1.4+ has (have not looked earlier) both a worker.add (singular) and a works.add (plural).  The call looks like it should be for workers.add and when I modify things to make this happen it stops erroring out but does not actually spin up the GPUs.

**Comment by Kettoch** [comment]
Yep, It seems that it does not like having the multiple workers added at one time.  I think it CAN, but you have to include the per algorithm parameters in the command.
something like this...
{"id":1,"method":"worker.add","params":["equihash","0","M=1","equihash","1","M=1"]}

And from the looks of it every algorithm has a different set of parameters.

If you want to start messing with them, I can definitely test for you.

**Comment by Kettoch** [comment]
Reverted back to the original settings and examined the call to the miner and this looks odd.

Here is the list of commands being passed to the miner.  Something looks odd.

`{
    "time": 1,
    "commands": [
        {
            "id": 1,
            "method": "worker.add",
            "params": [
                "equihash",
                "0",
                "equihash",
                "1"
            ]
        },
        {
            "id": 1,
            "method": "worker.add",
            "params": [
                "equihash",
                "0",
                "equihash",
                "1"
            ]
        }
    ]
}`

**Comment by Kettoch** [comment]
I have one that is working now, but I do not know how to propose a change here.

Line 82 in the minerNHMP file was

`[PSCustomObject]@{time = 1; commands = @(@([PSCustomObject]@{id = 1; method = "worker.add"; params = @($Miner_Device.PlatformId_Index | ForEach-Object {@("$Algorithm", "$_")} | Select-Object)}) * $Threads) + $Params}`

This line seems to have worked but would be interested to see if any 3 or 4 nvidia rigs work with this.

`[PSCustomObject]@{time = 1; commands = @(for ($i=0;$i -le ($Threads-1);$i++){$Miner_Device.PlatformId_Index | ForEach-Object {@([PSCustomObject]@{id = 1; method = "worker.add"; params = @(@("$Algorithm", "$_") | Select-Object)}) + $Params}})}`

**Comment by Kettoch** [comment]
Now, when the miner is idle, I get this. (after this change)

`[[2018-06-16 08:56:14.741987] [thread=0x00002a7c] [info]]

	Connected to nhmp.usa.nicehash.com:3200


[[2018-06-16 08:56:14.820113] [thread=0x000075d4] [info]]

	nhmp | Subscribed


[[2018-06-16 08:56:19.236870] [thread=0x000075d4] [info]]

	nhmp | Disconnected


[[2018-06-16 08:56:19.238902] [thread=0x000075d4] [warning]]

	nhmp | read: End of file (2)


[[2018-06-16 08:56:29.434816] [thread=0x000075d4] [info]]

	Connected to nhmp.usa.nicehash.com:3200


[[2018-06-16 08:56:29.524789] [thread=0x00002a7c] [info]]

	nhmp | Subscribed


[[2018-06-16 08:56:33.917871] [thread=0x00002a7c] [info]]

	nhmp | Disconnected


[[2018-06-16 08:56:33.917871] [thread=0x00002a7c] [warning]]

	nhmp | read: End of file (2)


[[2018-06-16 08:56:44.068376] [thread=0x00002a7c] [info]]

	Connected to nhmp.usa.nicehash.com:3200


[[2018-06-16 08:56:44.218451] [thread=0x000075d4] [info]]

	nhmp | Subscribed


[[2018-06-16 08:56:48.554210] [thread=0x000075d4] [info]]

	nhmp | Disconnected


[[2018-06-16 08:56:48.616711] [thread=0x000075d4] [warning]]

	nhmp | read: End of file (2)


[[2018-06-16 08:56:58.802477] [thread=0x000075d4] [info]]

	Connected to nhmp.usa.nicehash.com:3200


[[2018-06-16 08:56:58.901953] [thread=0x00002a7c] [info]]

	nhmp | Subscribed


[[2018-06-16 08:57:03.301937] [thread=0x00002a7c] [info]]

	nhmp | Disconnected


[[2018-06-16 08:57:03.301937] [thread=0x00002a7c] [warning]]

	nhmp | read: End of file (2)


`

**Comment by UselessGuru** [comment]
@Kettoch 

Thank you for reporting back.

> I get the error when merging these files into mine.

```
Compare-Object : Cannot bind argument to parameter 'ReferenceObject' because it is null.
At C:\MultiPoolMiner\MultiPoolMiner.ps1:532 char:129
```

This error has nothing to do with Excavator and has since fixed. See 
https://github.com/MultiPoolMiner/MultiPoolMiner/commit/4eee162154cca37aa83a014df8ab857b84b82dbd
https://github.com/MultiPoolMiner/MultiPoolMiner/pull/1867

Regarding worker.add & workers.add

This miner file does 1 or two instances (one per configured thread) of worker.add. There is no need for workers.add (it would just make the code more complex).

Now, when the miner is idle, I get this. (after this change)

`[[2018-06-16 08:56:14.741987] [thread=0x00002a7c] [info]]

Connected to nhmp.usa.nicehash.com:3200`

This is fine. Excavator runs as a service.

**Comment by Dewbs** [comment]
Tried 1.5.5a - same benchmark fault as previous report

**Comment by Kettoch** [comment]
@UselessGuru 
Your original call to the miner results in a fatal error from the binary stating an invalid stoi (string to integer?) call.

What I did is change the call to create additional single worker.add commands.

Look at my comment about line 82 (and its corresponding line in the next set).

This is what the original line sends.
`api | Received: {"id":1,"method":"worker.add","params":["equihash","0","equihash","1"]}`

Results in an error. however, this:
`[[2018-06-17 04:55:08.549972] [thread=0x0000cb78] [trace]]
	api | Received: {"id":1,"method":"worker.add","params":["lyra2rev2","0"]}
[[2018-06-17 04:55:09.741478] [thread=0x0000cb78] [trace]]
	api | Received: {"id":1,"method":"worker.add","params":["lyra2rev2","1"]}
`
This does work.  And the change I made corrected that.  Is there something from my side that you want in order for you to see what it is I am referring to?  Because I can assure you that before I made those two line changes, the newer excavator never ran.  The binary simply fatally errored out with an STOI error in the command.

**Comment by UselessGuru** [comment]
> It lists miner in each gpu section but then assigns collective rate to each gpu.

I don't quite follow. Can you please post a screenshot, the contents of http://localhost:3999/devices as well as the relevant excavator log?

**Comment by UselessGuru** [comment]
@Kettoch 

Can you please post the contents of http://localhost:3999/devices as well as the relevant excavator log?

**Comment by Kettoch** [comment]
Here you go:
[MultiPoolMiner_2018-06-17.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2109482/MultiPoolMiner_2018-06-17.txt)
[Excavator-5400_2018-06-17_18-37-14.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2109485/Excavator-5400_2018-06-17_18-37-14.txt)
[devices.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2109490/devices.txt)

**Comment by Dewbs** [comment]
Let’s say card A makes 15h/s, card B makes 20, card C makes 25.  There is a total rate of 60. 
60 gets assigned to card A, card B & card C after benchmark.

**Comment by UselessGuru** [comment]
@Dewbs
Please post the contents of http://localhost:3999/devices as well as the relevant excavator log.

I try to figure you why your config starts mining, whereas Kettoch's  bombs out.

**Comment by Dewbs** [comment]
[Excavator-5400_2018-06-17_20-51-01.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2112344/Excavator-5400_2018-06-17_20-51-01.txt)
[Devices.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2112347/Devices.txt)

**Comment by UselessGuru** [comment]
@Kettoch 

Thank you for your input. I've modified the miner file - hopefully this works for you too!

**Comment by Dewbs** [comment]
Doesn't change incorrect bench
[Excavator-5400_2018-06-18_19-12-45.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2112463/Excavator-5400_2018-06-18_19-12-45.txt)

**Comment by Dewbs** [comment]
Still same.
I'm not sure you are clear of my issue?

**Comment by UselessGuru** [comment]
> I'm not sure you are clear of my issue?

Yes, I am... your box has 3 different cards, and the current ExcavatorNHMP API does not handle this yet.... I am working on it.

**Comment by Kettoch** [comment]
@UselessGuru This is no longer failing but it is not creating worker.add lines for card number 1, only for card number 0.

[Excavator-5400_2018-06-18_16-07-25.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2112845/Excavator-5400_2018-06-18_16-07-25.txt)

The problem with the line is that you need two increments.  One provides the number of worker.add threads for a GPU and an outside incremental that creates that set for each GPU.
2 threads 4 cards gets 8 worker.adds.

What the current one seems to do is give you the right amount of threads but ALL reference the first card.

**Comment by Dewbs** [comment]
That’s great, will test as soon as I can (6hrs or so)

**Comment by Dewbs** [comment]
Seems to be working.
Was a strange bench period with certain cards sitting idle for complete cycles, but mostly completed I think.

**Comment by Dewbs** [comment]
[Excavator-5400_2018-06-19_20-10-51.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116654/Excavator-5400_2018-06-19_20-10-51.txt)
[Excavator-5400_2018-06-19_20-10-44.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116655/Excavator-5400_2018-06-19_20-10-44.txt)
[Excavator-5400_2018-06-19_20-10-12.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116656/Excavator-5400_2018-06-19_20-10-12.txt)
[Excavator-5400_2018-06-19_20-09-27.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116657/Excavator-5400_2018-06-19_20-09-27.txt)
[Excavator-5400_2018-06-19_20-09-20.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116658/Excavator-5400_2018-06-19_20-09-20.txt)
[Excavator-5400_2018-06-19_20-08-49.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116659/Excavator-5400_2018-06-19_20-08-49.txt)
[Excavator-5400_2018-06-19_20-08-03.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116660/Excavator-5400_2018-06-19_20-08-03.txt)
[Excavator-5400_2018-06-19_20-07-56.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116661/Excavator-5400_2018-06-19_20-07-56.txt)
[Excavator-5400_2018-06-19_20-07-28.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2116662/Excavator-5400_2018-06-19_20-07-28.txt)

**Comment by Kettoch** [comment]
@UselessGuru Thank you for the changes.  It appears to be running all of the workers it should be.  I am verifying logs and already have a new message but it looks more annoying than anything else.  I am one of those programmers that hates yellow and reds messages, can't help it.

However...
https://github.com/MultiPoolMiner/MultiPoolMiner/commit/4eee162154cca37aa83a014df8ab857b84b82dbd
https://github.com/MultiPoolMiner/MultiPoolMiner/pull/1845/commits/2884dea737b642b98abe749f9ef441f100a1a90e
https://github.com/MultiPoolMiner/MultiPoolMiner/pull/1845/commits/72d69ed6e18b303dc4950d72a8519a3d3d398071

Were all not integrated to my configuration.

The message I get is this:
`[18:32:57][0x00008f04][info] nhmp | Disconnected
[18:32:57][0x00008f04][warning] nhmp | read: The I/O operation has been aborted because of either a thread exit or an application request (995)
[18:32:59][0x00008f04][info] Connected to nhmp.usa.nicehash.com:3200
[18:32:59][0x00008ce4][info] nhmp | Subscribed
[18:33:01][0x00008f04][info] nhmp | daggerhashimoto | add success`

**Comment by Kettoch** [comment]
What do those MinMem settings mean?  I think one of them might be set to high.  Are there cards with 12GB of RAM in the NVIDIA line (if my math is right)

**Comment by UselessGuru** [comment]
@Kettoch 
> What do those MinMem settings mean? I think one of them might be set to high. 


That is on purpose as not to run miners if there is not enought memory available.
e.g. Neoscrypt in 2 threads will not work with 11GB... 
If you happen to know more appropriate memory requirements or see that some algorithms still run out of memory then please report it here.

> read: The I/O operation has been aborted because of either a thread exit or an application request (995)

This is when Excavator unsubscribes from the stratum server where there are no more workers.

**Comment by UselessGuru** [comment]
@Dewbs 

I am getting myself another NVidia card so I can hopefully reproduce your issue.

**Comment by Kettoch** [comment]
@UselessGuru  Now I wish looking up your profile had a BTC address on it... :)

**Comment by UselessGuru** [comment]
@Kettoch 

If you search you'll find it... and you'll get disappointed :-)

**Pull request closed without merging**